module/zfs/zap.c

```diff
@@ -1189,6 +1219,21 @@ fzap_cursor_retrieve(zap_t *zap, zap_cursor_t *zc, zap_attribute_t *za)
 	/* retrieve the next entry at or after zc_hash/zc_cd */
 	/* if no entry, return ENOENT */
 
+	/*
+	 * If we are reading from the beginning, we're almost certain to
+	 * iterate over the entire ZAP object.  If there are multiple leaf
+	 * blocks (freeblk > 2), prefetch the whole object (up to
+	 * dmu_prefetch_max bytes), so that we read the leaf blocks
+	 * concurrently. (Unless noprefetch was requested via
+	 * zap_cursor_init_noprefetch()).
+	 */
+	if (zc->zc_hash == 0 && zap_iterate_prefetch &&
+	    zc->zc_prefetch && zap_f_phys(zap)->zap_freeblk > 2) {
+		dmu_prefetch(zc->zc_objset, zc->zc_zapobj, 0, 0,
```

Comment:
That's all correct. `dmu_prefetch()` doesn't know about the ZAP format, it's just fetching the specified byte range of the object (capped by `dmu_prefetch_max`).  Basically we are assuming that large (>128MB) ZAP objects are rare enough that their performance is not important enough to add the more complicated logic here to prefetch incrementally in zap hash order.  If we find this is not the case, my first choice would be to increase the 128MB limit rather than adding that logic :-)